File: big data(3 project)/k-FoldCrossValidation.R
# Διασταυρωμένη Επικύρωση k-Πτυχών


#
# Συνάρτηση που υπολογίζει και επιστρέφει το Μέσο Τετραγωνικό Σφάλμα (Root Mean Squared Error 
# - RMSE) 
# predictedValues: διάνυσμα με τιμές της εξαρτημένης μεταβλητής που προβλέπει το μοντέλο 
#                  παλινδρόμησης
# actualValues: διάνυσμα με τις πραγματικές τιμές της εξαρτημένης με-ταβλητής
# Επιστρέφει το Μέσο Τετραγωνικό Σφάλμα πρόβλεψης
calculateRMSE<-function(predictedValues, actualValues){
  err<- sqrt( mean((actualValues - predictedValues)^2)  )
  return( err )
}


# Συνάρτηση που υλοποιεί τον αλγόριθμο της διασταρωμένης επικύρωσης k-πτυχών.
# Η συνάρτηση κάνει χρήση του μέσου τετραγωνικού σφάλματος (RMSE) ως μετρική σφάλματος.
# Παράμετροι συνάρτησης:
# data: το σύνολο δεδομένων που θα χωριστεί σε τμήματα ελέγχουν κα εκπαίδευσης
# frml: το γραμμικό μοντέλο παλινδρόμησης που θα αξιολογηθεί η ακρί-βεια πρόβλεψής του
# k: η τιμή k της διασταυρωμένης επικύρωσης k-πτυχών που δηλώνει σε πόσα τμήματα θα διαχωριστεί
#    το αρχικό σύνολο δεδομένων
# Επιστρεφόμενη τιμή:
# Η συνάρτηση επιστρέφει τον μέσο όρο του μέσου τετραγωνικού σφάλματος
kFoldCrossValidation<-function(data, frml, k){
  # Τυχαία αναφιάταξη των παρατηρήσεων του συνόλου δεδομένων
  dataset<-data[sample(nrow(data)),]
  #Δημιουργία k σε πλήθος τμημάτων του συνόλου δεδομένων με περίπου ίσο πλήθος 
  # παρατηρήσεων σε κάθε τμήμα.
  folds <- cut(seq(1,nrow(dataset)), breaks=k, labels=FALSE)
  # Διάνυσμα όπου αποθηκεύεται το Μέσο Τε
  RMSE<-vector()
  # Επαναλαηπτική διαδικασία όπου κάθε ένα από τα k τμήματα θα χρησι-μοποιηθεί διαδοχικά
  # ως σύνολο ελέγχου για το μοντέλο παλινδρόμησης και όλα τα υπόλοιπα ως σύνολο εκπαίδευσης. 
  # Η διαδικασία θα τερματίσει εάν όλα τα τμήματα έχουν χρησιμοποιηθεί ως σύνολο σλέγχου.
  for(i in 1:k){
    # Καθορισμός του τμήματος ελέγχου για την τρέχουσα επανάληψη 
    testIndexes <- which(folds==i,arr.ind=TRUE)
    # Καθορισμός συνόλου ελέγχου μοντέλου
    testData <- dataset[testIndexes, ]
    # Καθορισμός συνόλου εκπαίδευσης μοντέλου, που θα είναι όλα τα υπόλοιπα
    # πλην των δεδομένων που χρησιμοποιηθούν για έλεγχο
    trainData <- dataset[-testIndexes, ]
    # Εκτίμηση συντελεστών του μοντέλου παλινδρόμησης χρησιμοποιώντας το σύνολο εκπαίδευσης
    candidate.linear.model<-lm( frml, data = trainData)
    # Υπολογισμός των τιμών της εξαρτημένης μεταβλητής που προβλέπει το μοντέλο 
    # για τις τιμές του τρέχοντος συνόλου ελέγχου 
    predicted<-predict(candidate.linear.model, testData)
    # Υπολογισμός σφάλματος RMSE
    error<-calculateRMSE(predicted, testData[, "mpg"])
    # Αποθήκευση τιμής σφάλματος
    RMSE<-c(RMSE, error)
  }
  # Επιστροφή μέσης τιμής των σφαλμάτων που προέκυψαν απ'όλα τα τμήμα-τα ελέγχου
  return( mean(RMSE) )
}


# Ανάγνωση δεδομένων
# Το αρχείο δεδομένων περιέχει τεχνικά χαρακτηριστικά αυτοκινήτων διαφόρων κατασκευαστών
carData<-read.csv("auto-mpg.csv", sep=";", header=T, stringsAsFactors = F, quote = "\"")


# Έλεγχος για δεδομένα που λείπουν (missing values) και αφαίρεση ολόκληρης γραμμής
# σε περίπτωση που μια τουλάχιστον μεταβλητή έχει missing value
carData<-na.omit(carData)

# Επειδή οι μεταβλητές horsepower και mpg είχαν missing values, πρέπει αυτές να 
# μετατραπούν στον σωστό τύπο δεδομένων (numeric), καθότι είχαν αναγνωστεί ως συμβολοσειρές 
# (character) εξαιτίας των missing value
carData[,"horsepower"] <- as.numeric(carData[,"horsepower"])
carData[,"mpg"] <- as.numeric(carData[,"mpg"])

# Τα τέσσερα υποψήφια μοντέλα των οποίων θα αξιολογηθεί η ικανότητα πρόβλεψης της κατανάλωσης καυσίμου
# (mpg - miles per gallon) με τη μέθοδο της διασταυρωτικής επικύρωσης 10-φορές. 
# Τα μοντέλα παλινδρόμησης αποθηκεύονται στο διάνυσμα ως συμβολοσειρές και θα μετατραπούν
# σε τύπους (formula) της R

predictionModels<-vector()
predictionModels[1]<-"mpg ~ horsepower+weight"
predictionModels[2]<-"mpg ~ horsepower+acceleration+displacement"
predictionModels[3]<-"mpg ~ horsepower+displacement+weight"

# Συμπεριλαμβάνεται και μοντέλο που εισάγει πολυωνυμικό όρο βαθμού 2 weight^2. Η εισαγωγή τέτοιων
# όρων στο μοντέλο απαιτεί τη χρήση της συνάρτησης I() που είναι η συνάρτηση Inhibit Interpretation και
# έχει σαν αποτέλεσμα να μην ερμηνευτεί ο τελεστής ^ στα πλαίσια του τύπου.
# Αυτό γιατί ο τελεστής ^ έχει ειδκή σημασία για τύπους και αν χρησι-μοποιείται σε τέτοιους δίχως τη
# χρήση της I() δεν θα ερμηνευτεί ως ο τελεστής ύψωση σε δύναμη.
predictionModels[4]<-"mpg ~ horsepower+displacement+weight + I(weight^2)"

# Μετά από κάθε διασταυρωμένη επικύρωση, ο μέσος όρος του μέσου τετρα-γωνικού 
# σφάλματος κάθε μοντέλου θα αποθηκευτεί σε διάνυσμα.
modelMeanRMSE<-vector()

# Διασταυρωμένη επικύρωση 10-φορές για κάθε ένα από τα 
# τέσσερα υποψήφια μοντέλα.
for (k in 1:length(predictionModels)){
  # Δισταυρωμένη επικύρωση 10-πτυχών για το γραμμικό μοντέλο παλιν-δρόμησης k
  modelErr<-kFoldCrossValidation(carData, as.formula(predictionModels[k]), 10)
  # Αποθήκευση του μέσου σφάλματος
  modelMeanRMSE<-c(modelMeanRMSE, modelErr)
  print( sprintf("Linear regression model [%s]: prediction error [%f]", predictionModels[k], modelErr ) )
}



# Ποιο μοντέλο είχε το χαμηλότερο μέσο τετραγωνικό σφλάμα;
bestModelIndex<-which( modelMeanRMSE == min(modelMeanRMSE) )
# Εμφάνιση μοντέλου με το μικρότερο μέσο τετραγωνικό σφάμα δηλαδή τη μεγαλύτερη ακρίβεια

print( sprintf("Model with best accuracy was: [%s] error: [%f]", predictionModels[bestModelIndex], modelMeanRMSE[bestModelIndex]) )

# Για το μοντέλο με το χαμηλότερο μέσο σφάλμα, εκτιμώνται οι συντελε-στές του 
# λαμβάνοντας υπόψη ολόκληρο το σύνολο δεδομένων ως σύνολο εκπαίδευσης
final.linear.model<-lm( as.formula(predictionModels[bestModelIndex]), data=carData )

File: big data(3 project)/K-FoldCrossValidation.py
from sklearn.linear_model import LinearRegression
from math import sqrt # We'll need sqrt()
import statistics # for mean()
from sklearn.metrics import mean_squared_error  
from sklearn.model_selection import KFold 
from sklearn.preprocessing import PolynomialFeatures
import numpy as np
import pandas as pd


# Read the data
hhData = pd.read_csv("HouseholdData.csv", header=0, sep=",", engine='python')

# Randomly shuffle the data i.e. mix it up randomly.
# We do this in order to get different partitions during k-fold cross validation between
# different executions of the program.
# This isn't required per se, but we do it for educational purposes
hhData = hhData.sample(frac=1).reset_index(drop=True)

# We will execute k-fold cross validation to assess the regression model with respect to
# its accuracy to predict the value of the dependent variable.
# The linear regression model that will be assessed is the following:
# FoodExpenditure = b1Income + b2FamilySize + b0
# We will assess its accurace in predicting the dependent value using k-fold cross validation


# First, we setup the k-fold cross validation object.
# We will do a 5-fold cross validation model (i.e. k=5) i.e. the dataset will
# be split in 5 partitions with an approximately equal number of observation in each part.
# We use the KFold object from sklearn and initialize it properly
kf = KFold(n_splits=5) 


print("\nLinear regression model: FoodExpenditure = b1Income + b2FamilySize + b0\n")

# Create an empty array where we will store the calculated RMSE values
# so that we may be able to 
allRMSE = np.empty(shape=[0, 1])

# Just a variable to count at which test we currently are
testNumber = 0

# Start now iterating over the partitioned dataset, selecting each time a different subset as the testing set.
#
# .split(hhData) will split the dataset into 5 parts.
# Now we iterate over these parts. This iteration works as follows:
# variables train_index and test_index will get the indexes of the original dataset (hhData)
# that will constitute the training- and testing-set respectively.
# This for loop will be executed 5 times, equal to the number of partitions.
for train_index, test_index in kf.split(hhData):

 # Next test
 testNumber += 1
 
 # Use the current indexes train_index and test_index to get the actual observations for the
 # training and testing of the model respectively.

 # From the original, complete dataset get the data with which we will use TRAIN our model (aka the training set),
 # i.e. estimate the coefficients. We get the rows designated by train_index and all their columns/variables
 trainingData = hhData.iloc[train_index,:]

 # From the original, complete dataset get the data with which we will TEST our model, i.e. estimate its
 # accuracy. This is the "unknown" dataset.
 # Note: we do know the value of the dependent variable FoodExpenditure for the testing set
 # and hence we will be able to estimate the prediction error/accuracy
 testData = hhData.iloc[test_index,:]

 # Use the training data to estimate the coefficients of the multiple linear regression model.
 # The model we will estimate is: FoodExpenditure = b1Income + b2FamilySize + b0
 lm = LinearRegression(fit_intercept=True)

 # Since the linear regression model has as independent variables Income and FamilySize,
 # we get these variables from the training dataset.
 # The method .fit() does the actual estimation of the coefficients for the linear regression
 # model using OLS (Ordinary Least Squares).
 # The first argument of the .fit() method are the values of the independent variables -in our case
 # trainingData.loc[:,['Income','FamilySize']] - and the second
 # argument are the values of the dependent variable, here trainingData.loc[:,['FoodExpenditure']].
 estimatedModel = lm.fit(trainingData.loc[:,['Income','FamilySize']], trainingData.loc[:,['FoodExpenditure']])

 # Coefficients have been estimated. Take a look at them. Not that it is important but heck, why not.
 # NOTE: the coefficients are returned (.coef_) as an array. The two numbers you'll see when
 # printing .coef_  must be interpreted as follows: first number is the coefficient
 # for Income, second number the coefficient for FamilySize. The estimated constant term b0 can retrieved 
 # via the .intercept_ variable.
 print(">>>Iteration ", testNumber, sep='')
 print("\tEstimated coefficients:")
 print("\t\tb1=", estimatedModel.coef_[0][0] , sep='')
 print("\t\tb2=", estimatedModel.coef_[0][1] , sep='') 
 print("\t\tb0=", estimatedModel.intercept_, sep='')
 
 
 # Now, use the estimated model to predict the value for FoodExpenditure for the observations in the testing set
 # (i.e. the unknown data that was not used for estimating the coefficients).
 # For this, we'll use the model's .predict() method which takes as argument the values of the independent variables
 # in the linear regression model.
 # IMPORTANT NOTE: Since we gave to the .fit() method above the the variables in a specific order (Income and FamilySize) we
 # have to give the respective variables of the testing set in the same order.
 # The .predict() method will return a vector of predicted values, one for each observation in testData.loc[:,['Income','FamilySize']].
 # More specifically, the first value in the vector predictedExpenditure is the predicted value for FoodExpenditure for the first row in
 # testData.loc[:,['Income','FamilySize']], the second value is the predicted value for FoodExpenditure for the second row
 # testData.loc[:,['Income','FamilySize']] etc. 
 predictedExpenditure = estimatedModel.predict(testData.loc[:,['Income','FamilySize']])

 # Calculate the Root Mean Squared Error (RMSE) for this testing set.
 # We have the real values of the dependent variable from the original dataset
 # which is testData.loc[:,['FoodExpenditure']] and the model's predicted values in predictedExpenditure
 # IMPORTANT: Please note the following: the function mean_squared_error() calculate the MSE NOT the RMSE. In
 # order to get the RMSE, we need to square the value returned by mean_squared_error(). See your notes on
 # how MSE and RMSE differ.
 RMSE = sqrt(mean_squared_error(testData.loc[:,['FoodExpenditure']], predictedExpenditure))

 # Display the RMSE value
 print("\t\tModel RMSE=", RMSE, sep='')
 
 # Also, store the calculated RMSE value into an array, so that we can calculate the mean error after k-fold cross
 # validation has been finished 
 allRMSE = np.append(allRMSE, RMSE)

 # Here the for-loop ends and restarts again for a different training- and testing set.





# Ok. The iterations of k-fold cross validation have been done.
#
# We have now 5 values of RMSE, one for each testing set.
# We calculate the mean RMSE which gives a better estimate on how accurate the predictions
# of the linear regression model is for unknown data and thus
# a better estimate for the generalization error.
print("\n=======================================================")
print(" Final result: Mean RMSE of tests:", statistics.mean(allRMSE), sep='' )
print("=======================================================")


